src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,360 @@
+             PreparedStatement defaultInsert = conn.prepareStatement(commands.insertOrIgnoreCommand(
+                     "INSERT ", "INTO pendingusage (user_id, writer_id, pending_bytes) VALUES(?, ?, ?)"));
+             PreparedStatement insert = conn.prepareStatement("UPDATE pendingusage SET pending_bytes = pending_bytes + ? " +
+                     "WHERE writer_id = ?;")) {
+            try {
+                defaultInsert.setInt(1, userId);
+                defaultInsert.setInt(2, writerId);
+                defaultInsert.setInt(3, 0);
+                defaultInsert.executeUpdate();
+
+                insert.setLong(1, size);
+                insert.setInt(2, writerId);
+                int count = insert.executeUpdate();
+                if (count != 1)
+                    throw new IllegalStateException("Didn't update one record!");
+                conn.commit();
+            } catch (SQLException e) {
+                conn.rollback();
+                throw e;
+            }
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public UserUsage getUsage(String username) {
+        int userId = getUserId(username);
+        try (Connection conn = getNonCommittingConnection();
+             PreparedStatement search = conn.prepareStatement("SELECT pu.writer_id, pu.pending_bytes, uu.total_bytes, uu.errored " +
+                "FROM userusage uu, pendingusage pu WHERE uu.user_id = pu.user_id AND uu.user_id = ?;");
+             PreparedStatement writerSearch = conn.prepareStatement("SELECT key_hash FROM writers WHERE id = ?;")) {
+            try {
+                search.setInt(1, userId);
+                ResultSet resultSet = search.executeQuery();
+                Map<PublicKeyHash, Long> pending = new HashMap<>();
+                long totalBytes = -1;
+                boolean errored = false;
+                while (resultSet.next()) {
+                    writerSearch.setInt(1, resultSet.getInt(1));
+                    ResultSet writerRes = writerSearch.executeQuery();
+                    writerRes.next();
+                    PublicKeyHash writer = PublicKeyHash.decode(writerRes.getBytes(1));
+                    pending.put(writer, resultSet.getLong(2));
+                    if (totalBytes == -1) {
+                        totalBytes = resultSet.getLong(3);
+                        errored = resultSet.getBoolean(4);
+                    }
+                }
+                conn.commit();
+                return new UserUsage(totalBytes, errored, pending);
+            } catch (SQLException e) {
+                conn.rollback();
+                throw e;
+            }
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getUserId(String username) {
```

Comment:
could pass in the 'live' connection to private methods like this.

Reply:
That might be faster, but I think this is okay for now.

Reply:
My only thought is that if it is another connection it may not 'see' the changes made in the calling method. Not sure if new connection inherits transaction or is separate..

Reply:
If it doesn't see it then it would fail, because the userid constraint would fail. I think.

Reply:
ok I've inlined the calls to this and getwriterid which are between sql statements